File: 2021/src/day_10.rs
pub const TITLE: &str = "--- Day 10: Syntax Scoring ---";

use std::fs;

pub fn part_1() -> i64 {
    let openers: [char; 4] = ['(', '{', '<', '['];
    let closers: [char; 4] = [')', '}', '>', ']'];
    let closer_points: [i64; 4] = [3, 1197, 25137, 57];

    fs::read_to_string("inputs/day10")
        .expect("Unable to read input")
        .lines()
        .map(|line| {
            let mut stack: Vec<char> = Vec::new();
            let mut error_score = 0;
            for bracket in line.chars() {
                if openers.contains(&bracket) {
                    stack.push(bracket);
                } else {
                    let score = match stack.pop() {
                        Some(x) => {
                            let opener_index = openers.iter().position(|&o| o == x).unwrap();
                            let closer_index = closers.iter().position(|&c| c == bracket).unwrap();
                            if opener_index == closer_index {
                                0
                            } else {
                                closer_points[closer_index]
                            }
                        }
                        None => 0,
                    };
                    error_score += score;
                }
            }
            error_score
        })
        .sum()
}

pub fn part_2() -> i64 {
    let openers: [char; 4] = ['(', '{', '<', '['];
    let closers: [char; 4] = [')', '}', '>', ']'];
    let auto_complete_points: [i64; 4] = [1, 3, 4, 2];

    let mut scores: Vec<i64> = fs::read_to_string("inputs/day10")
        .expect("Unable to read input")
        .lines()
        .filter_map(|line| {
            line.chars()
                .fold(Some(Vec::new()), |stack, bracket| match stack {
                    Some(mut stack) => {
                        if openers.contains(&bracket) {
                            stack.push(bracket);
                            Some(stack)
                        } else {
                            match stack.pop() {
                                Some(opener) => {
                                    let opener_index =
                                        openers.iter().position(|&o| o == opener).unwrap();
                                    let closer_index =
                                        closers.iter().position(|&c| c == bracket).unwrap();
                                    if opener_index == closer_index {
                                        Some(stack)
                                    } else {
                                        None
                                    }
                                }
                                None => Some(stack),
                            }
                        }
                    }
                    None => None,
                })
        })
        .map(|stack| {
            stack.iter().rev().fold(0, |score, bracket| {
                let opener_index = openers.iter().position(|&o| o == *bracket).unwrap();
                (score * 5) + auto_complete_points[opener_index]
            })
        })
        .collect();

    scores.sort();
    scores[scores.len() / 2]
}

File: 2021/python/day_10.py
def part_1():
    opening = ["(", "{", "<", "["]
    closing = [")", "}", ">", "]"]
    illegal_points = {")": 3, "]": 57, "}": 1197, ">": 25137}

    stack, error_score = [], 0
    with open("inputs/day10", "r") as file:
        for line in file.read().splitlines():
            for char in line:
                if char in opening:
                    stack.append(char)
                elif char in closing:
                    if len(stack) == 0:
                        continue
                    if opening.index(stack.pop()) != closing.index(char):
                        error_score += illegal_points[char]

    return error_score


def part_2():
    opening = ["(", "{", "<", "["]
    closing = [")", "}", ">", "]"]
    auto_complete_points = {")": 1, "]": 2, "}": 3, ">": 4}

    auto_complete_scores = []
    with open("inputs/day10", "r") as file:
        for line in file.read().splitlines():
            corrupted, stack = False, []
            for char in line:
                if char in opening:
                    stack.append(char)
                elif char in closing:
                    if len(stack) == 0:
                        continue
                    if opening.index(stack.pop()) != closing.index(char):
                        corrupted = True
                        break

            if corrupted:
                continue
            current_score = 0
            for char in stack[::-1]:
                current_score = (current_score * 5) + auto_complete_points[
                    closing[opening.index(char)]
                ]
            auto_complete_scores.append(current_score)

    auto_complete_scores.sort()
    return auto_complete_scores[len(auto_complete_scores) // 2]


print(part_1())
print(part_2())
